### ISSUE TITLE
amp-bind: Expression complexity limit reached

### ISSUE THREAD

**Issue by Gigliocom** [ampproject/amphtml#11434]
## What's the issue?

We need to perform over 50 expressions in unique AMP page, how can we do it?

## How do we reproduce the issue?

Below a demo AMP product page with the errors.
https://www.giglio.com/borse-donna_borsa-mano-borsa-petite-2-jours-piccola-con-barra-metallica-logo-inciso-fendi-8bh2533wl.html?cSel=027&AAMMPP

## What browsers are affected?

All browsers

## Which AMP version is affected?

V0

**Comment by camelburrito** [comment]
@Gigliocom - Thanks for reaching out! Could you provide detailed steps on how to reproduce the issue (and what this issue is). 

I did not see any errors on the link above.

**Comment by Gigliocom** [comment]
Hi,
when you click on thumb to choose another color, the main photo must be change and the hidden values into form must change also. But that does not.

In console.log appear this:
error.js:169 amp-bind: Expression eval failed with error: Expression size (52) exceeds max (50). Please reduce number of operands.​​​
Uncaught (in promise) Error: amp-bind: Expression eval failed with error: Expression size (52) exceeds max (50). Please reduce number of operands.​​​ _reported_

**Comment by choumx** [comment]
Hi Giglio.com, thanks for your feedback. The expression complexity limit is intended to keep usages of `amp-bind` fast and responsive. 

One way you can simplify your page expressions (and speed up the page itself!) is to reduce the number of variables. Specifically, I think you can remove the `selectedSlideFor*` variables. 

Instead of showing/hiding 4 different `<amp-carousel>` elements based on the selected color, try using a single `<amp-carousel>` with its `<amp-img>` children changing their `src` via a `[src]` binding. For example:

```html
<amp-state id="product">
  {
    "027": {
      "image1" : "images/027/1.jpg",
      "image2": "images/027/2.jpg",
       ...
    }
  }
</amp-state>

<amp-carousel>
  <amp-img [src]="product[selectedColor].image1"></amp-img> 
  <amp-img [src]="product[selectedColor].image2"></amp-img> 
</amp-carousel>
```

Hope that helps!

**Comment by Gigliocom** [comment]
Hi @choumx,
thanks for your help, but i think that this is not solution. We have products that has many photo per colors but unfortunately any colors dont'have the same number of pictures.

There is an another way to execute more than 50 expressions?

**Comment by choumx** [comment]
> We have products that has many photo per colors but unfortunately any colors dont'have the same number of pictures.

Ah, yes. We're planning to support more dynamic slide content in `amp-carousel` (#8921). One temporary workaround is to create a carousel with the max number of images among your products, and hide or replace overflow images with a stock image.

> There is an another way to execute more than 50 expressions?

Not at the moment, though we may revisit this if it's a common problem -- I'll use this issue to track. Intermediary variables would also help reduce duplication of expression fragments in your case (#8397). 

Sorry that you're have difficulties with `amp-bind`. We'll try to improve this soon.

**Comment by ampprojectbot** [comment]
This issue doesn't have a category which makes it harder for us to keep track of it. @choumx Please add an appropriate category.

**Comment by ivanliu8** [comment]
Definitely need to increase the limit. I'm creating a page with personalised products. Customer have options to choose designs, colours, pictures, fonts and enter custom text. We are using amp-bind to build up a preview link. 50 operands are definitely not enough. Alternatively, is there any ways I can encode the state json object and pass to the server as an parameter?

**Comment by choumx** [comment]
@ivanliu8 Can you share a sample page with expressions that exceed this limit?

**Comment by leonalicious** [comment]
may i know how to calculate / mean by 50 operands ? we did a page, we didnt change any things on frontend, but we did change backend by adding 1 more layer and it shows me this exceed 50 operands error.
@choumx

**Comment by leonalicious** [comment]
may i know how do you calculate the operands ? such as AstNodeType.NOT is equivalent to how many operands , AstNodeType.TERNARY is equivalent to how many operands, a reference to the source code would be much appreciated. Thx in advanced ! @choumx

**Comment by choumx** [comment]
"Operand" is loosely defined here. 😄  It's technically the number of nodes in the abstract syntax tree of the expression.

I'd really appreciate any example pages that exceed the operand limit. This way we can thoughtfully support your use cases without necessarily changing the performance characteristics of `amp-bind`.

**Comment by leonalicious** [comment]
a work around trick would be:

based on the documentation, `input` would support both `tap` and `change` actions. So, basically you can create a **transparent dummy input** in front of your `<element>`

a simple demo as below: 
1. set the parent `position: relative`
`<input type="checkbox" class="dummy-checkbox"
 on="tap: AMP.setState({
 50  operands bla bla bla
});
change: AMP.setState({
 50 operands bla bla bla
})"
>`

css:
.dummy-checkbox {
    opacity: 0;
    position: absolute;
    width: 100%;
    height: 100%;
    z-index: 1;
}

**Comment by leonalicious** [comment]
a work around trick would be:

based on the documentation, `input` would support both `tap` and `change` actions. So, basically you can create a **transparent dummy input** in front of your `<element>`

a simple demo as below: 
1. set the parent `position: relative`

    <input type="checkbox" class="dummy-checkbox" on="
    tap: AMP.setState({
        50  operands bla bla bla
    });
    change: AMP.setState({
        50 operands bla bla bla 
    })">

    css:
    .dummy-checkbox {
        opacity: 0;
        position: absolute;
        width: 100%;
        height: 100%;
        z-index: 1;
    }

**Comment by leonalicious** [comment]
a work around trick would be:

based on the documentation, `input` would support both `tap` and `change` actions. So, basically you can create a **transparent dummy input** in front of your `<element>`

a simple demo as below: 

1. set the parent `position: relative`

code:

    <input type="checkbox" class="dummy-checkbox" on="
    tap: AMP.setState({
        50  operands bla bla bla
    });
    change: AMP.setState({
        50 operands bla bla bla 
    })">

    css:
    .dummy-checkbox {
        opacity: 0;
        position: absolute;
        width: 100%;
        height: 100%;
        z-index: 1;
    }

**Comment by gtonysam** [comment]
HI ,

I am facing the same issue, I try to pass say 10 or more state parameters in url, but it is failing 
amp-bind: Expression compile error in Expression size (61) exceeds max (50). Please reduce number of operands.

**Comment by choumx** [comment]
@gtonysam Which component URL are you mutating? `amp-list` or `amp-state`?

**Comment by gtonysam** [comment]
@choumx  Hi we are using amp-state to bind the values to the url.
We have as many as 20 paramters to pass to the url. We tried amp-bind macro as well to add values but still the below issue comes always- amp-bind :Expression compile error in Expression size () exceeds max (50). Please reduce number of operands.

**Comment by choumx** [comment]
Can you share the markup for this element?

**Comment by gtonysam** [comment]
Hi Choumx,

This is what I am trying to do,

 <amp-bind-macro id="dynamicDate" arguments="" expression="searchForm.departureDay+searchForm.departureMonthURL+searchForm.departureMonth+searchForm.departureYearURL" />
    <amp-bind-macro id="LocationURL" arguments="" expression="searchForm.originLocationURL+searchForm.source+searchForm.destinationLocationURL+searchForm.destination" />
    
    <amp-bind-macro id="dynamicURL" arguments="" expression="searchForm.baseurl+searchForm.tripType+dynamicDate()+LocationURL()+searchForm.endURL" />


This is just a part of url and few more operands needs to be added and I am unable to add one more operand as it exceed the limit

**Comment by choumx** [comment]
Where is `dynamicURL` used in your case? 

It looks like most of these variables are from `<input>` fields of a form. Have you considered flipping the logic? E.g. use amp-bind to modify the values of these inputs (e.g. `[value]` binding) and submit the form when ready?

**Comment by gtonysam** [comment]
dynamicURL is the base url, the searchform parameters are to be passed in the url. But the url is not a straightforward one as it contains string values.

Example of sample url : https://example.com/&outboundOption.departureDay=+"**SearchForm.startDate**"+/&outboundOption.departureMonth=+"**SearchForm.startMonth**"

The searchform values are dynamic values. We used amp-macro-bind , still the size limit exceeds. So we need to append multiple string values as well as input values

**Comment by gtonysam** [comment]
@choumx Any updates please ?

**Comment by choumx** [comment]
Does my suggestion above work for you? I'd need to see the entire page markup to fully understand what would or wouldn't work.

**Comment by gtonysam** [comment]
@choumx  Please provide you email ID - so that I can share the entire page itself, so that you will be able to understand what the exact issue is.

**Comment by choumx** [comment]
You can reach on privately on the amphtml Slack channel with the same username.

**Comment by gtonysam** [comment]
[amp-macros-06Feb old - Copy.txt](https://github.com/ampproject/amphtml/files/1710049/amp-macros-06Feb.old.-.Copy.txt)
 
Attached it text format

**Comment by gtonysam** [comment]
@choumx  is it possible to call REST service in amp ?

**Comment by choumx** [comment]
Sorry busy at AMP Conf and haven't gotten a chance to look at your example yet. 

For amp-bind use cases, you can try using the `src` attribute on `amp-state` which loads data from a JSON endpoint.

**Comment by choumx** [comment]
@gtonysam Suggested an alternate implementation here: https://github.com/ampproject/amphtml/issues/13429#issuecomment-367798919

**Comment by leonalicious** [comment]
@gtonysam you may use `amp-list` to call REST service as `amp-state` will not fetch upon initial page load.

**Comment by gtonysam** [comment]
@choumx  @leonalicious  Thanks for your suggestions. I have called REST service using amp-list to form the url string. But I am facing a new issue as the client side validation is not working proper;y. Even though it is fired, the click event is still triggering.

**Comment by choumx** [comment]
> But I am facing a new issue as the client side validation is not working proper;y. Even though it is fired, the click event is still triggering.

Not sure what this means, can you elaborate?

**Comment by gtonysam** [comment]
@choumx  Apologies for late message. I have suppressed the Bind issue using Amp-list rest service call. On button click, the parameters are passed to REST service and actual url is formed at the service and on button click, the url is called. I have used client side validation here. On tap I have used showerrors:true, eventhough the validation fails, the url is being called

**Comment by choumx** [comment]
@gtonysam Glad to hear you were able to solve the issue. Sounds like you're having an issue with `amp-form` client-side validation?

**Comment by gtonysam** [comment]
@choumx  I am not using amp-form, but yes I am facing client side validation. Is there a way to perform client side validation in amp-list.

**Comment by choumx** [comment]
@gtonysam I'd highly recommend using `amp-form` for client-side validation. `amp-bind` can also hook into the validation events. See details here: https://www.ampproject.org/docs/reference/components/amp-form#custom-validations

**Comment by gtonysam** [comment]
@choumx  As I mentioned,I am using Amp-list and used Get request to fetch the  url from rest service. I have seen amp-form used for POST . Will this still work ?

**Comment by Makac** [comment]
I definitely agree that the expression complexity limit should be increased.

In my current project, hitting the complexity limit forces me to instead do frequent server round-trips to process data, that could easily be done client-side with a higher limit. This definitely makes the page more sluggish than using expressions would. So in this case the complexity limit is actually counteractive to the the reason behind it (to keep pages fast and responsive).

Another way to adress this (because there will always be cases where amp expressions won't be able to do the job for you), would be to have the ability to do state updates via an iframe. Basically you'd feed input data to the iframe, which in turn processes the data and updates the amp-state. 

On a side note, it would be very useful to be able to “stringify” an amp-state object or a portion of it, because I frequently end up in situations where I'd like to pass the state (or parts of it) in a server request. (Attempts to do this with expressions is a good example of how you hit the complexity limit, by the way). It definitely also be useful for aforementioned iframe communication.

**Comment by weeblrpress** [comment]
Hi

@choumx I am also hit by the limit in what I thought was a straightforward use case. Here is the markup:

````<div` class="wbamp-comment-status wbamp-error i-amphtml-error" [hidden]="(avlVar.current_pa_color == 'green' && avlVar.current_size == 'XL') || (avlVar.current_pa_color == 'green' && avlVar.current_size == 'M') || (avlVar.current_pa_color == 'blue' && avlVar.current_size == 'XL') || (avlVar.current_pa_color == 'blue' && avlVar.current_size == 'L') || (avlVar.current_pa_color == 'blue' && avlVar.current_size == 'M')">
			Sorry, this product is unavailable. Please choose a different combination.        </div>````

1 - AMP runtime says this has 59 operands, which is very different to what we can evaluate as "operand" by looking at the JS code
2 - Any workaround (short of splitting the condition and repeating the error message N times)?

Best regards

**Comment by leonalicious** [comment]
@weeblrpress this may solve your issue

**Comment by choumx** [comment]
@Makac Thanks for the feedback. We're considering loosening this restriction, though a concrete use case or example that's currently blocked would help out a lot!

@weeblrpress Another way to do it is store the available color/size lookup in an `amp-state`. This should  make your expressions more concise and readable. E.g.

```html
<amp-state id="inStock">
  <script type="application/json">
  {
    "green": {
      "M": true,
      "L": false,
    },
    ...
  }
  </script>
</amp-state>

<div [hidden]="inStock[avlVar.current_pa_color][avlVar.current_size]"></div>
```

**Comment by weeblrpress** [comment]
HI

@leonalicious Thanks for your suggestion. Your solution can indeed improve things in this specific instance, but I'm afraid it won't scale that much when there are more than a few attributes, more options for an attribute or the displays conditions are not as simple.

@choumx Yes, that sounds like the way to go,and I'm already using amp-state on this type of pages, but it does require to rewrite some code to validate all the details. I'll post back.

Thanks

**Comment by Makac** [comment]
@choumx, I'd be happy to provide detailed info on what it is I'm trying to do. Is there a private channel I can use for this?

**Comment by choumx** [comment]
@Makac Great. You can message me privately on the AMP Slack channel, same username. https://github.com/ampproject/amphtml/blob/master/CONTRIBUTING.md#discussion-channels

**Comment by choumx** [comment]
> On a side note, it would be very useful to be able to “stringify” an amp-state object or a portion of it, because I frequently end up in situations where I'd like to pass the state (or parts of it) in a server request. (Attempts to do this with expressions is a good example of how you hit the complexity limit, by the way). It definitely also be useful for aforementioned iframe communication.

@Makac Thanks for the feature request. We could do this by adding a new URL replacement variable that reads from `amp-state` data, e.g. `<amp-list src="https://data.com?foo=STATE(foo)&foobar=STATE(foo.bar)">`. Though perhaps an easier way is to stop counting dereferences as an operand.

> Currently sitting at 6 over the 50 limit. I'm building a filterable inventory list that gets its JSON data from our ERP systems API. I'm trying to find ways to reduce the operands in my on change event but not seeing it yet - still working on it though. My second amp-list (Machine Make) is the one that is maxing out on me.

@craigscott29 Thanks for posting your use case. I think it looks reasonable.

Since people are hitting the "50 operand" limit much more often than the "2000 binding" limit, I'm going to see if we can shift this balance to something like 100 operands and 1000 bindings and still retain (roughly) the same performance bounds. 

Thanks for your patience everyone.

**Comment by craigscott29** [comment]
@choumx any chance of the operand increase having a quick turn around? The president of the company I work for wants the inventory list live in the next 2 weeks haha.

**Comment by craigscott29** [comment]
@choumx, thanks. I shall take a look into it. I had tried something like in my first attempt by following [this](https://ampbyexample.com/samples_templates/product_browse_page/) example but I couldn't seem to get the filtering to work. Variables were passing correctly to the query string, it just wasn't doing anything.

**Comment by choumx** [comment]
It won't do anything automatically. That sample shows server-side filtering/sorting -- the idea is to read the query parameters on the server and then perform the sorting on the server. 

OTOH, your current code performs filtering/sorting on the client. This is fine for simple examples but can hit complexity limits as you've seen. :) Generally, AMP strives to "solve problems at the right layer", so we believe that more complex business logic should be done on the server, not the client.

Of course there are shades of gray in between -- sometimes it's nice to not perform a network request when changing a filter. For your case, since you're performing an network request anyways, doing the filtering on the server should work. Hope that makes sense!

**Comment by craigscott29** [comment]
I may be thinking about this incorrectly, but I think the server-side is where my issue may be. The data I'm retrieving doesn't exist on the same server as the site. I'm scraping it from our ERP systems API which is data output only. Essentially, I can retrieve my block of data in one pull (which is all I'd want to do because it's a slow system) and can't query against the server itself. 

I'll maybe try setting up a service to pull the data down daily to the web server.

**Comment by choumx** [comment]
Ah, I see. Yea, having a flexible server is very useful when developing with AMP in general.

**Comment by zoezl** [comment]
Hi @choumx 


I'm trying to render an expression with amp-list and facing the same issue.

This is for calculating the price of merchandise with different additions.
The number of elements in attr array and content array is uncertain. I have no idea how this could work within the limit.

```
<amp-list src="json/prodetails.json" layout="fill" id="list_id">
    <template type="amp-mustache">
      <p [text]="
        {{#sliver}}
        {{#attr}}
          materialData.items[0].slive[0].attr[{{id}}].content[pro.{{name}}Id?(+pro.{{name}}Id)+1:0].price
        {{/attr}}
        {{/sliver}}
      "></p>
    </template>
  </amp-list>
```
Is there any workaround?

**Comment by leonalicious** [comment]
Since people are hitting the "50 operand" limit much more often than the "2000 binding" limit, I'm going to see if we can shift this balance to something like 100 operands and 1000 bindings and still retain (roughly) the same performance bounds.
To be frank, we are not really sure how to do the counting for the bindings, currently we got 2 concerns about reducing the "2000" bindings limit down to "1000" :

We do not know how the "garbage collector disposal" logic works, but as long as it works like a charm, I think very rare case it will hit the bindings limit. We did encounter 2000 bindings limit issue before, please refer to https://github.com/ampproject/amphtml/issues/12517, and based on my wild guess, the root cause was <amp-state [src]>, it seems like the garbage collector disposal logics did not work well with <amp-state [src]>.
There are use cases that new items are being appended to <amp-list> upon user clicks on "View More" button and what the action does is to submit a <form> and keep accumulating the data [src] of the list. Our concern is will this action cause the bindings limit issue if users click on "View More" button a lot of times ?
While for the decision of increasing the operands from 50 to 100, we are quite happy with it, but it may lead to another question. FYI, due to the "less logic restriction" of mustache template, we are writing most of our if...else... logic with <div [class]="true ? 'show' : 'hide'" approach, and if we expand the operands from 50 to 100, I can foresee there will be some blinkings performance issue, if some developers write some very complex hide & show method on the <div> elements.

**Comment by choumx** [comment]
Update: In the last 7 days, there have been < 50 instances where an AMP page exceeds 1000 bindings (including runtime changes e.g. "view more"  button) across very few domains. 

I think we can safely rebalance the complexity limit to 100 operands and 1000 bindings, which should help unblock many use cases. I'll open another issue to address the "view more" and infinite scroll use case.

**Comment by choumx** [comment]
Fixed by #15517.

**Comment by JakobEichler** [comment]
I'd like to execute this assignment, but I get this error:

<form id="show-more-form-default-search" method="POST" action-xhr="<?php echo $srcSearch?>" on="submit-success: AMP.setState( {
            searchMeta: { offset: searchMeta.offset + 10 },
            defaultParkingPlaces: {
                my_response: [ {
                    items: defaultParkingPlaces.my_response[0].items.concat(  event.response.my_response[0].items),
                    resultsCount: event.response.my_response[0].resultsCount,
                    has_results: event.response.my_response[0].has_search_results,
                    has_more_results: event.response.my_response[0].has_more_results,
                    timeSpanStr: event.response.my_response[0].timeSpanStr,
                    placeName: event.response.my_response[0].placeName
                } ],
            }
        } );">



amp-bind: Expression eval failed. Expression size (104) exceeds max (100). Please reduce number of operands.​​​
Error: amp-bind: Expression eval failed. Expression size (104) exceeds max (100). Please reduce number of operands.​​​


I made my response so nested, because I wanted support for global variables in an amp-list template where I also use the same endpoint.

**Comment by choumx** [comment]
Can you post an example of this large expression please?

**Comment by JakobEichler** [comment]
`<form id="show-more-form-default-search" method="POST" action-xhr="<?php echo $srcSearch?>" on="submit-success: AMP.setState( {
            searchMeta: { offset: searchMeta.offset + <?php echo $limitShowMore?> },
            defaultParkingPlaces: {
                my_response: [ {
                    items: defaultParkingPlaces.my_response[0].items.concat( event.response.items),
                    resultsCount: event.response.my_response[0].resultsCount,
                    has_results: event.response.my_response[0].has_results,
                    has_more_results: event.response.my_response[0].has_more_results,
                    timeSpanStr: event.response.my_response[0].timeSpanStr,
                    placeName: event.response.my_response[0].placeName
                } ],
            }
        } );">`



But I changed my response and removed the my_response[0]. part and it works now.
